### PR TITLE
[SPARK-15083] [Web UI] History Server can OOM due to unlimited TaskUIData

### PR DESCRIPTION
**Pull request by ajbozarth**
## What changes were proposed in this pull request?

Based on #12990 by @tankkyo

Since the History Server currently loads all application's data it can OOM if too many applications have a significant task count. `spark.ui.trimTasks` (default: false) can be set to true to trim tasks by `spark.ui.retainedTasks` (default: 10000) 

(This is a "quick fix" to help those running into the problem until a update of how the history server loads app data can be done)
## How was this patch tested?

Manual testing and dev/run-tests

spark-15083

### CODE REVIEW

**Comment by ajbozarth** [comment]
Found a odd behavior while testing. When a application encounters an issue (in my case my task queue filled up and SparkListenerBus was stopped) it can cause the tasks HashMap size to be larger than the sum of completed, failed, and active tasks (total tasks). This means that the UI can end up looking like the following two screen shots (of the same application). If everyone can agree this is a separate issue from this I will file a JIRA for it for later.

with `spark.ui.trimTasks` true:
screen shot 2016-08-16 at 2 29 07 pm

with `spark.ui.trimTasks` false:
screen shot 2016-08-16 at 2 20 40 pm

**Comment by ajbozarth** [comment]
Pushed a update to the text so it's clearer in case of the above bug:
screen shot 2016-08-16 at 2 47 28 pm
screen shot 2016-08-16 at 2 48 30 pm

**Comment by ajbozarth** [comment]
@vanzin and @tgravescs here's my update of pr #12990 by @tankkyo if you wan to take a look, I raised a few discussion points above

**Comment by tgravescs** [comment]
yes the queue filling up and dropping events is a separate issue there are other jiras filed for it.

File: core/src/main/scala/org/apache/spark/ui/SparkUI.scala [hunk 1]
```diff
@@ -141,6 +141,7 @@ private[spark] object SparkUI {
   val DEFAULT_POOL_NAME = "default"
   val DEFAULT_RETAINED_STAGES = 1000
   val DEFAULT_RETAINED_JOBS = 1000
+  val DEFAULT_RETAINED_TASKS = 10000
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/ui/SparkUI.scala]: do we have an idea of when people are hitting this?  Personally I find 10000 small also, I was thinking more like 50000 or 100000 and users can limit further if they have the history server configured to use small amounts of memory.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala [hunk 2]
```diff
@@ -93,6 +93,8 @@ class JobProgressListener(conf: SparkConf) extends SparkListener with Logging {
 
   val retainedStages = conf.getInt("spark.ui.retainedStages", SparkUI.DEFAULT_RETAINED_STAGES)
   val retainedJobs = conf.getInt("spark.ui.retainedJobs", SparkUI.DEFAULT_RETAINED_JOBS)
+  val retainedTasks = conf.getInt("spark.ui.retainedTasks", SparkUI.DEFAULT_RETAINED_TASKS)
+  val trimTasks = conf.getBoolean("spark.ui.trimTasks", false)
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: I don't think we need the on/off switch for the trimTasks.  None of the others have off switches so I say just leave it without for now and a user can set to something huge like 100,000,000.  We can always add an off later if needed (like set to -1). If we are worried about an int being to small we could make it a long.  I don't see anything have over 2 billions tasks at this point though.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala [hunk 3]
```diff
@@ -93,6 +93,8 @@ class JobProgressListener(conf: SparkConf) extends SparkListener with Logging {
 
   val retainedStages = conf.getInt("spark.ui.retainedStages", SparkUI.DEFAULT_RETAINED_STAGES)
   val retainedJobs = conf.getInt("spark.ui.retainedJobs", SparkUI.DEFAULT_RETAINED_JOBS)
+  val retainedTasks = conf.getInt("spark.ui.retainedTasks", SparkUI.DEFAULT_RETAINED_TASKS)
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: ideally we use the new config builder stuff

Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: actually looking at the taskId's they are a long so we should make this a long.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala [hunk 4]
```diff
@@ -137,6 +139,17 @@ class JobProgressListener(conf: SparkConf) extends SparkListener with Logging {
     )
   }
 
+  /** If Tasks is too large, remove and garbage collect old tasks */
+  private def trimTasksIfNecessary(taskData: HashMap[Long, TaskUIData]) = synchronized {
+    if (taskData.size > retainedTasks) {
+      val toRemove = (taskData.size - retainedTasks)
+      val oldIds = taskData.map(_._2.taskInfo.taskId).toList.sorted.take(toRemove)
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: is there a reason we are using _._2.taskInfo.taskId instead of just the Long ( _._1), which is the taskId also?

I'm a bit concerned with the sorting here.  If someone is using really large number of tasks , say 1 million I'm afraid this sorting might be expensive, especially if we are doing it for every single task that comes in after hitting the limit, we have to keep resorting every time.

File: core/src/main/scala/org/apache/spark/ui/SparkUI.scala (hunk 1 above)
Comment (ajbozarth) [inline core/src/main/scala/org/apache/spark/ui/SparkUI.scala]: I'll change it to 100000

File: core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala (hunk 2 above)
Comment (ajbozarth) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: Ok, that works for me

File: core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala (hunk 3 above)
Comment (ajbozarth) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: I'm not sure I understand why this should be a long, also I haven't used the new config builder yet, should I update all of the vals to use it?

File: core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala (hunk 4 above)
Comment (ajbozarth) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: This code is from the original pr, I didn't know _._1 is always equal to _._2.taskInfo.taskId so I'll change that, as for the loop, it does slow things down, I noticed it in my testing, I'll look at a better way to do this.

File: core/src/main/scala/org/apache/spark/ui/SparkUI.scala (hunk 1 above)
Comment (ajbozarth) [inline core/src/main/scala/org/apache/spark/ui/SparkUI.scala]: As for the why, the issue is when the history server is full of applications with large task amounts, there is no issue with just one application having a significant number of tasks by itself.

**Comment by ajbozarth** [comment]
@tgravescs I've addressed some issues: I changed the default max to 100000 and removed the on/off. I also updated the `taskData` to be a `LinkedHashMap` instead of a `HashMap` which allows for easier trimming without losing or changing the current functionality. I'll be pushing the code later today once I finish my tests.

**Comment by ajbozarth** [comment]
ran into a odd test failure in HistoryServer I still haven't figured out, I'll push an update to fix it tomorrow

File: core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala (hunk 3 above)
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: a long would just allow it to be set to a higher value.  2^64 vs 2^32.  Since the TaskId is a long already (see TaskInfo.scala) it makes sense to make the config a long. It really only matters if someone wants to set the limit over 2 billionish.

We have been trying to make any new configs the ConfigEntry type. so you don't have to do them all just the new one.  Its pretty straight forward, take a look at https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/internal/config/package.scala

Comment (ajbozarth) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: `retainedTasks` is compared to `stageData.taskData.size`, which is a `int`, I don't see how `TaskId` being a long is relevant. 

And I've switched to a `ConfigEntry` and will push my changes after testing.

Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: ah good point that hashMap can only hold int number of entries.  

My point was that with taskId being a long and assuming taskId = 0 through X, X here could be a long (2^64) and thus you could have 2^64 number of tasks.  If you can have 2^64 number of tasks then this config should allow you to configure it to retain all of those tasks.

I'm fine with leaving an int.

Comment (ajbozarth) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: Ah, now I get where you where going with that, yeah if it wasn't for the map size issue that would have been good.

**Comment by ajbozarth** [comment]
Pushed my latest updates including the change to using `ConfigEntry` and updating the test cases (using `LinkedHashMap` changes the secondary sort to `taskId`)

**Comment by tgravescs** [comment]
code changes look good, does it change the secondary sort of taskId in a bad way?  Looking at your json files it looks like they were in order before and now they aren't.

**Comment by ajbozarth** [comment]
secondary sort in this case is whatever order the collection is stored in, since `LinkedHashMap` keeps an order (essentially by `tasked`, but really by insertion order) `tasked` becomes the defacto secondary sort. In `HashMap` there's no guaranteed order and it's secondary sort is whatever order the hashes are stored (this can be consistent in controlled environments like test suites). 

tl;dr it doesn't change secondary sort in a bad way, it just provides an order that previously didn't exist.

**Comment by ajbozarth** [comment]
@srowen @vanzin or @JoshRosen could one of you take a look at this? I've already addressed @tgravescs feedback, and would love to see this go through soon if it looks good to you, thanks

**Comment by tgravescs** [comment]
sorry @ajbozarth somehow my mail didn't show me you updated it.

**Comment by tgravescs** [comment]
+1, I'll merge shortly

**Comment by tgravescs** [comment]
I merged this into 2.1.0, if you are wanting this in older ones can you put up PR for those, cherry-pick to branch-2.0 wasn't clean.

**Comment by ajbozarth** [comment]
@tgravescs would would I go about opening a pr for older branches? I've only done work on the latest code before. Also is the 1.6.x branch still updating or is 2.0.x and 2.1 the only updating branches now?

**Comment by tgravescs** [comment]
when you do the pr to apache/spark just pick branch-2.0 instead of master.  I see things going into branch-1.6 also but I haven't seen any discussion on the list about doing another release so we would have to ask there.

**Comment by ajbozarth** [comment]
ok, thanks, I'll open those pr's once I can cleanly merge the patch

**Comment by srowen** [comment]
You would also have to check out branch-2.0 and branch from there in order to work off the right base. From there you can cherry-pick your commit from master and resolve conflicts. Then there's nothing extra to do when opening a PR as it's already known to be vs branch-2.0. It's more than just choosing a branch when you open the PR.

**Comment by srowen** [comment]
Ah forgot to comment on the substance: normally we probably wouldn't back-port a feature to a maintenance branch. This is pretty much a feature. If it's low-risk and someone wants to make the case that it is important enough, it's possible. For example, you could argue that the potential for an OOM is a bug. My gut says it would be reasonable for 2.0.x

**Comment by ajbozarth** [comment]
Thanks @srowen based on that I'll just open it to the 2.0 branch then
